src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,142 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> firstNode = new Node<>(null, value, null);
+            head = firstNode;
+            tail = firstNode;
+            size++;
+            return;
+        }
+        tail.next = new Node<>(tail, value, null);
+        tail = tail.next;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index >= 0 && size == 0 || index == size) {
+            add(value);
+            return;
+        }
+        Node<T> currentNodeByIndex = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(currentNodeByIndex.prev, value, currentNodeByIndex);
+        currentNodeByIndex.prev = newNode;
+        if (index == 0) {
+            head = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        T oldItemInNode = nodeByIndex.item;
+        nodeByIndex.item = value;
+        return oldItemInNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T removedItem = findNodeByIndex(index).item;
+        unlink(findNodeByIndex(index));
+        return removedItem;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.item == object || (object != null && object.equals(currentNode.item))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
```

Comment:
We have at least two index checks - mb we need separate method for index check?

Reply:
In the first case, the index is checked for the next steps of the add method. In the second, the index is checked for valid values. So there is no sence in introducing a new method for checking the index